icu4j/main/core/src/main/java/com/ibm/icu/util/IslamicCalendar.java

```diff
@@ -732,6 +780,33 @@ static final double moonAge(long time)
     // Calendar framework
     //----------------------------------------------------------------------
 
+    /**
+     * The handleGetMonthLength function for Islamic Civil calendar.
+     */
+    private int handleGetMonthLengthIslamicCivil(int extendedYear, int month) {
+        return ((month+1) % 2) +
+            ((month == DHU_AL_HIJJAH && civilLeapYear(extendedYear)) ? 30 : 29);
+    }
+
+    /**
+     * The handleGetMonthLength function for Islamic calendar.
+     */
+    private int handleGetMonthLengthIslamic(int extendedYear, int month) {
+        month += 12*(extendedYear-1);
+        return (int)( trueMonthStart(month+1) - trueMonthStart(month) );
+    }
+
+    /**
+     * The handleGetMonthLength function for Islamic Umalqura calendar.
+     */
+    private int handleGetMonthLengthIslamicUmalqura(int extendedYear, int month) {
+        if (extendedYear < UMALQURA_YEAR_START || extendedYear > UMALQURA_YEAR_END) {
+            return handleGetMonthLengthIslamicCivil(extendedYear, month);
+        }
+        return ((UMALQURA_MONTHLENGTH[extendedYear - UMALQURA_YEAR_START] & // calculate year offset into bit map array
+              (0x01 << (11 - month))) == 0 ) ? 29 : 30; // set mask for bit corresponding to month
```

Comment:
This logic might also be worth exploding out into its constituent pieces, more or less the way it was before this refactoring.